### DISCUSSION TITLE
How to get the CalendarDatePicker related Textbox

### DISCUSSION THREAD

**Discussion by bsdRuf** [AvaloniaUI/Avalonia#13464] [Q&A] [answered]
I am trying to implement shorthands for the CalendarDatePicker. Like 3112 + Enter (or leave event) = 31/12/1999

But the only events I can use are for CalendarDatePicker itself, not the Textbox.

Is there such a thing like a DOM i can navigate through? If so, how?

Since I have multiple CalendarDatePicker I cant use the Textbox name 'PART_TextBox' to access them.

![image](https://github.com/AvaloniaUI/Avalonia/assets/136796815/b72c748c-f711-4d0b-bd70-1c3e87104a1a)

**Comment by robloo** [comment]
> Is there such a thing like a DOM i can navigate through? If so, how?

Yes, there are two: The Visual Tree and the Logical Tree. You want the Visual Tree in this case which includes all controls used to construct the UI.

You can use a helper function like this one: https://reference.avaloniaui.net/api/Avalonia.VisualTree/VisualExtensions/A64F5E09. Otherwise, you can navigate the tree recursively yourself.

You will also probably need to derive a new, customized version of the CalendarDatePicker control that you can then use everywhere.

---

All of that said, this seems like a good feature to have built into the control. Certain regions or applications have short-hand date notations that the app can handle better than the framework.

I propose adding a new property to the CalendarDatePicker that allows an app to set a Func<DateTime, string> that can be used to customize how a string is parsed and converted into a DateTime. That concept is new for a XAML framework as it wasn't available when all these controls were first written I think. However, something similar is being done in the AutoCompleteBox already.

**Reply by bsdRuf** [reply]
Thank you for the quick and well explained reply.

I tried this but its losing it straight after the DataValidationErrors 
```
        private TextBox getCPDTextbox(CalendarDatePicker cdp)
        {
            DataValidationErrors dve        = cdp.FindAncestorOfType<DataValidationErrors>();
            DockPanel dockP                 = dve.FindAncestorOfType<DockPanel>();
            ContentPresenter contentPres    = dockP.FindAncestorOfType<ContentPresenter>();
            Panel panel                     = contentPres.FindAncestorOfType<Panel>();
            Grid grid                       = panel.FindAncestorOfType< Grid> ();
            TextBox resTextBox              = grid.FindAncestorOfType<TextBox>();

            return resTextBox;
        }
```

Did I do anything wrong there?

Thanks in advance

**Reply by workgroupengineering** [reply, marked as answer]
You can only use it after the visual tree has been completely loaded

```c#
        private TextBox? getCPDTextbox(CalendarDatePicker cdp) => 
            cdp.GetVisualDescendants()
                .OfType<TextBox>()
                .FirstOrDefault(t => t.Name == "PART_TextBox");
``` 

see AttachedToVisualTree or Loaded events.

**Reply by timunie** [reply]
You can simply do

```cs
calendarDatePicker.GetTemplateChildren().OfType<TextBox>().First();

// or
calendarDatePicker.GetVisualDescendants().OfType<TextBox>().First()

// or
calendarDatePicker.FindDescendantOfType<TextBox>()
```
